MRN-Code/coinstac
# Redesign/rethink acl

* **Problem:**
   1. The redesign database solution requires authentication and an ACL plan that will limit user access to certain routes.
* **Solution:**
   1. Add a `users` table in the `coinstac` database to hold new users and their permissions.
   1. Additional `hapi` routes outside of GraphQL to handle authentication and account creation.
   1. Use `hapi-auth-jwt2` to generate token for authenticated users.
   1. User permissions determine front-end access and back-end GraphQL access.
* **Tests:**
   1. Follow the steps outlined in `packages/coinstac-api-server/README.md` to setup your development environment with RethinkDB, start the GraphQL API, and insert a couple computations into the new database.
   1. Duplicate `packages/coinstac-graphql/config/api-config-sample.json` as `api-config.json`.
   1. Uncomment the contents and fill in a random string for the `jwtSecret`.
   1. Duplicate `packages/coinstac-graphql/config/comp-reg-config-sample.json` as `comp-reg-config.json` and uncomment the contents.
   1. Start COINSTAC UI dev environment as usual. 
   1. Start the application and login with user: `test` & pass: `password` to test authentication.
   1. Notice that there is no route listed called "Submit Computation". That is because this user does not have write access.
   1. Refresh the application and login with a random user. An error should appear.
   1. Click on the Create User tab and create a new user using `test` as the username. You should see an error telling you that the username is already in use.
   1. Create a new random user to test that user creation works.
   1. Refresh the application and login with user: `author` & password: `password`. Notice that you have a navigation item called Submit Computation.
